src/CloudEvent.php

```diff
@@ -0,0 +1,123 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\CloudFunctions;
+
+class CloudEvent
+{
+    // Required Fields
+    private $id;
+    private $source;
+    private $specversion;
+    private $type;
+
+    // Optional Fields
+    private $datacontenttype;
+    private $dataschema;
+    private $subject;
+    private $time;
+    private $data;
+
+    public function __construct(
+        string $id,
+        string $source,
+        string $specversion,
+        string $type,
+        ?string $datacontenttype,
+        ?string $dataschema,
+        ?string $subject,
+        ?string $time,
+        ?array $data
+    ) {
+        $this->id = $id;
+        $this->source = $source;
+        $this->specversion = $specversion;
+        $this->type = $type;
+        $this->datacontenttype = $datacontenttype;
+        $this->dataschema = $dataschema;
+        $this->subject = $subject;
+        $this->time = $time;
+        $this->data = $data;
+    }
+
+    public function getId(): string
+    {
+        return $this->id;
+    }
+    public function getSource(): string
+    {
+        return $this->source;
+    }
+    public function getSpecVersion(): string
+    {
+        return $this->specversion;
+    }
+    public function getType(): string
+    {
+        return $this->type;
+    }
+    public function getDataContentType(): ?string
+    {
+        return $this->datacontenttype;
+    }
+    public function getDataSchema(): ?string
+    {
+        return $this->dataschema;
+    }
+    public function getSubject(): ?string
+    {
+        return $this->subject;
+    }
+    public function getTime(): ?string
+    {
+        return $this->time;
+    }
+    public function getData(): ?array
+    {
+        return $this->data;
+    }
+
+    public static function fromArray(array $arr)
+    {
+        $argKeys = ['id', 'source', 'specversion', 'type', 'datacontenttype', 'dataschema', 'subject', 'time', 'data'];
+        $args = [];
+        foreach ($argKeys as $key) {
+            $args[] = $arr[$key] ?? null;
+        }
+
+        return new static(...$args);
+    }
+
+    public function __toString()
+    {
+        $data_as_json = json_encode($this->data);
+        $output = implode("\n", [
+            'CLOUDEVENT:',
+            "- id: $this->id",
+            "- source: $this->source",
+            "- specversion: $this->specversion",
+            "- type: $this->type",
+            "- datacontenttype: $this->datacontenttype",
+            "- dataschema: $this->dataschema",
+            "- subject: $this->subject",
+            "- time: $this->time",
+            "- data: $data_as_json",
```

Comment:
This data could potentially include user-specific data, and so we should treat it as potentially PII.

For debugging purposes, it makes sense. I suggest outputting only on something like a `FUNCTIONS_DEBUG=1` environment variable.

Reply:
So, implement something like this?

```php
$shouldDisplayData = '1' === getenv('FUNCTIONS_DEBUG');
$dataToDisplay =  $shouldDisplayData ? $data_as_json : '[DATA]';
```

Reply:
Every CloudEvent JSON payload has a to string function. This function is never used by the framework itself, but is useful for debugging.

We don't put any restriction with the http signature and it's payload, so I don't see why we need to add anything for this param.

Reply:
Logging CloudEvent metadata is good because that's examining information about what's been delivered. The payload itself is application data. I do not think we should provide a CloudEvent troubleshooting facility, even as simple as a toString(), that exposes application data. In any language.

If something goes wrong and troubleshooting is needed, logging the CloudEvent is the natural thing to do, and if that gets into production it could create a privacy breach. I'd rather see this exclude the data payload entirely and leave it to the developer to log if needed.

Reply:
Agreed - if the logging of application is strictly necessary, it can be done with `getData()`

Reply:
Yes. I've removed it.